### PR TITLE
improvement: Replace most Iterable uses with Collection on Transaction and KeyValueService classes

### PR DESCRIPTION
**Pull request by LucasIME**
## General
**Before this PR**:

Almost all the methods on Transaction and KeyValueService received their `rows` parameters as `Iterable<byte[]>` , but that in practice we were mostly passing around actual collections.

This lead to cases [like this](https://github.com/palantir/atlasdb/blob/bf6890433442e3e1661fb64a5ec87a7ffee50c23/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java#L410-L426) where we were using Iterables as if they were Collections, when that is not guaranteed by the type. Iterables don't need to be backed by collections and can be backed by just iterators, which can be exhausted once iterated, leaving the Iterable empty. In many places throughout the codebase, we didn't take that into consideration and were trying to double consuming iterators leading to potentially subtle bugs (or at least hard to reason about code).

So, in order to make things easier to reason about and actually match most of the use we actually have, I'm changing all of our method signatures that didn't return iterators or streams to received Collections instead of Iterables.

**After this PR**:

==COMMIT_MSG==
Transaction and KeyValueServices methods changed to received rows as Collections instead of Iterables, except when originally returning iterators or streams.
==COMMIT_MSG==

**Priority**:

**Concerns / possible downsides (what feedback would you like?)**:

**Is documentation needed?**:

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java [hunk 1]
```diff
@@ -1805,13 +1805,13 @@ public void close() {
...
     @Override
-    public void addGarbageCollectionSentinelValues(TableReference tableRef, Iterable<Cell> cells) {
+    public void addGarbageCollectionSentinelValues(TableReference tableRef, Collection<Cell> cells) {
         try {
             final Value value = Value.create(PtBytes.EMPTY_BYTE_ARRAY, Value.INVALID_VALUE_TIMESTAMP);
             cellValuePutter.putWithOverriddenTimestamps(
                     "addGarbageCollectionSentinelValues",
                     tableRef,
-                    Iterables.transform(cells, cell -> Maps.immutableEntry(cell, value)));
+                    cells.stream().map(cell -> Maps.immutableEntry(cell, value)).collect(Collectors.toList()));
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java:1814]: Do we have allocation & performance test results comparing before & after? I'm a little concerned about materializing the intermediate collection in places like this and we may need to use Guava lazy transforms instead:

```suggestion
                    Collections2.transform(cells, cell -> Maps.immutableEntry(cell, value)));
```

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/KeyValueServices.java [hunk 2]
```diff
@@ -208,7 +209,8 @@ public static RowColumnRangeIterator mergeGetRowsColumnRangeIntoSingleIterator(
         Map<byte[], RowColumnRangeIterator> rowsColumnRanges =
                 kvs.getRowsColumnRange(tableRef, rows, batchColumnRangeSelection, timestamp);
         // Return results in the same order as the provided rows.
-        Iterable<RowColumnRangeIterator> orderedRanges = Iterables.transform(rows, rowsColumnRanges::get);
+        Iterable<RowColumnRangeIterator> orderedRanges =
+                rows.stream().map(rowsColumnRanges::get).collect(Collectors.toList());
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/KeyValueServices.java:213]: This is allocation sensitive, may want to keep lazy transform

```suggestion
        Iterable<RowColumnRangeIterator> orderedRanges = Collections2.transform(rows, rowsColumnRanges::get);
```

File: atlasdb-service/src/main/java/com/palantir/atlasdb/jackson/TableRowSelectionDeserializer.java [hunk 3]
```diff
@@ -42,10 +44,19 @@ public TableRowSelection deserialize(JsonParser jp, DeserializationContext ctxt)
...
         }
     }
+
+    private static <T> Collection<T> iterableToCollection(Iterable<T> iterable) {
+        Collection<T> collection = new ArrayList<T>();
+        for (T item : iterable) {
+            collection.add(item);
+        }
+        return collection;
+    }
```
Comment (schlosna) [inline atlasdb-service/src/main/java/com/palantir/atlasdb/jackson/TableRowSelectionDeserializer.java:61]: I think we could avoid the full collection copy (or at least do a direct sized copy via ArrayList constructor) by making `AtlasDeserializers.deserializeRows` return `Collection<byte[]>` as `JsonNodeIterable` could determine the size of nodes using something like below.

```
    private static class JsonNodeCollection<T> extends AbstractCollection<T> implements Collection<T> {
        private final JsonNode node;
        private final java.util.function.Function<JsonNode, T> generator;
        private final int size;

        JsonNodeCollection(JsonNode node, java.util.function.Function<JsonNode, T> generator) {
            this.node = node;
            this.generator = generator;
            this.size = (node == null) ? 0 : node.size();
        }

        @Override
        public int size() {
            return size;
        }

        @Override
        public Iterator<T> iterator() {
            if (node == null) {
                return Collections.emptyIterator();
            }
            return new AbstractIterator<T>() {
                private int index = 0;

                @Override
                protected T computeNext() {
                    JsonNode subNode = node.get(index++);
                    if (subNode == null) {
                        return endOfData();
                    }
                    return generator.apply(subNode);
                }
            };
        }

        @Override
        public String toString() {
            return Iterables.toString(this);
        }
    }
```

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java [hunk 4]
```diff
@@ -479,14 +479,18 @@ public void testRangePagingBatches() {
...
                         TEST_TABLE,
-                        Iterables.limit(
-                                Iterables.cycle(
-                                        RangeRequest.builder().batchHint(1000).build()),
-                                100),
+                        requests,
+                        //                        Iterables.limit(
+                        //                                Iterables.cycle(
+                        //                                        RangeRequest.builder().batchHint(1000).build()),
+                        //                                100),
```
Comment (schlosna) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java:493]:
```suggestion
                        requests,
```

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java [hunk 5]
```diff
@@ -479,14 +479,18 @@ public void testRangePagingBatches() {
         for (int i = 0; i < totalPuts; i++) {
             putDirect("row" + i, "col1", "v1", 0);
         }
-
+        List<RangeRequest> requests = IntStream.range(0, 100)
+                .boxed()
+                .map(_x -> RangeRequest.builder().batchHint(1000).build())
```
Comment (schlosna) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java:484]: nit:

```suggestion
                .mapToObj(_x -> RangeRequest.builder().batchHint(1000).build())
```

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java [hunk 6]
```diff
@@ -502,15 +506,12 @@ public void testRangePagingBatchesReverse() {
             putDirect("row" + i, "col1", "v1", 0);
         }
 
+        List<RangeRequest> requests = IntStream.range(0, 100)
+                .boxed()
+                .map(_x -> RangeRequest.builder().batchHint(1000).build())
```
Comment (schlosna) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java:511]:
```suggestion
                .mapToObj(_x -> RangeRequest.builder().batchHint(1000).build())
```

**Review by schlosna — COMMENTED** [review]
Have we tested downstream impact of API break and potential perf impacts?

**Review by jeremyk-91 — COMMENTED** [review]
I'm on board with this change in terms of where it gets us, though there are a couple of traps we do need to be aware of - as @schlosna notes these would mainly be perf and compatibility.

- For perf: in general, this means that translations of things that consume `Iterable` should be lazy transforms rather than eager ones (@schlosna identified a couple of these). We don't have good benchmarks unfortunately but at least some kind of staged rollout might be nice 
- For compatibility: It might be good to cut an RC and try and patch atlasdb in large internal product and atlasdb proxy.

**Comment by mdaudali** [comment]
Closing out as we're not working on this in the near term

**Pull request closed without merging**